src/gui/qgsmapcanvas.cpp

```diff
@@ -202,6 +202,7 @@ QgsMapCanvas::QgsMapCanvas( QWidget * parent, const char *name )
   mCurrentLayer = NULL;
   mMapOverview = NULL;
   mMapTool = NULL;
+  mOldCursor = cursor();
```

Comment:
MapTools is used only in qgsmapcanvas and cursor is changed in mapcanvas
too, that is why the save and restore is done in qgsmapcanvas. Due to the
fact that maptools don’t restore cursor, the previous cursor shape is kept
in mOldCursor.

In the original code the problem was not only that cursor always is set to
Arrow when a MapTool is deactivated, but it was impossible to change it
even there is a slot for the signal emit mapToolSet( NULL );

For example you can catch this signal in your application in a slot,
outside of qgsmapcanvas and change the cursor shape if MapTool is NULL.
(MapTool is deactivated), but pay attention that after emitting of the
signals the cursor is set to Arrow. This means that after changing the
cursor in the slot, the cursor will be changed again in qgsmapcanvas after
emitting of the signals!

This means that there are 2 problems in qgsmapcanvas:

   -

   The old state of the cursor is not kept and restored when MapTool is
   deactivated.
   -

   The change of the cursor to Arrow is after emiting of the signals. Has
   to be before emitting of the signals.






On Tue, Jul 28, 2015 at 3:49 PM, Matthias Kuhn <notifications@github.com>
wrote:

> In src/gui/qgsmapcanvas.cpp
> <https://github.com/qgis/QGIS/pull/2224#discussion_r35642452>:
>
> > @@ -202,6 +202,7 @@ QgsMapCanvas::QgsMapCanvas( QWidget * parent, const char *name )
> >    mCurrentLayer = NULL;
> >    mMapOverview = NULL;
> >    mMapTool = NULL;
> > +  mOldCursor = cursor();
>
> If the cursor is changed within the map tool it is the map tools
> responsibility to restore it in the end.
>
> Is there a reason you would change the cursor in a slot connected to this
> signal. Actually, this could happen in any slot connected to any signal.
> How is this particular signal special?
>
> I wonder if we could just remove this line instead.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qgis/QGIS/pull/2224/files#r35642452>.
>